[RISCV] Vendor Relocations for Xqci extension

svs-quic opened the pull request:
This patch implements vendor relocation support for RISC-V, starting with the Xqci extensions.

Vendor Relocations need to emit the vendor symbol, and add an `R_RISCV_VENDOR` relocation against that symbol before the specific vendor relocation itself.

This patch adds a `maybeAddVendorReloc` function which is called from `addReloc`, to implement this functionality. Vendor identifier symbols are cached, so that at most one is emitted per vendor identifier. 

Vendor identifiers symbols do not interfere with identically named symbols used by assembly.

lenary commented:
Tests for the eld implementation are here: https://github.com/qualcomm/eld/pull/16

topperc commented on llvm/include/llvm/MC/MCAsmBackend.h:
```diff
@@ -125,6 +125,18 @@ class MCAsmBackend {
   }
 
+  /// Hook to add any marker relocations that the target may need in addition
+  /// to the normal relocations on instructions. Currently used only by RISCV
```
RISCV -> RISC-V

MaskRay commented on llvm/lib/MC/ELFObjectWriter.cpp:
```diff
@@ -1435,6 +1435,21 @@ void ELFObjectWriter::recordRelocation(MCAssembler &Asm,
+  unsigned PreRelocType;
+  MCSymbol *PreRelocSymbol;
+  uint64_t PreRelocAddend;
+  if (Backend.fixupNeedsMarkerELFRelocation(Asm, *Fragment, Fixup, PreRelocType,
```
(MCAssembler.cpp and ELFObjectWriter.cpp have significant tech debt. I try to clean up them from time to time. Just noticed that my latest change 19730e37ec5d3513d188a75999c964f2ef4b9509 might need you to rebase.)

I think we should refrain from adding new hooks. (I added `Backend.handleAddSubRelocations` in the past, and on reflection we might need some cleanup so that it could be combined with getRelocType).

Can you add an extra fixup when `fixup_riscv_qc_e_branch` is appended? Or the hook is here because in some scenarios the fixup can be resolved and you don't want to add a redundant marker?

lenary replied:
> I think we should refrain from adding new hooks.

I do too, but "refrain from" doesn't mean "never". We could not find another way to accomplish this.

> Or the hook is here because in some scenarios the fixup can be resolved and you don't want to add a redundant marker?

This. For instance, branches can be evaluated when not relaxing.

From my comment, above:
> We didn't want a separate fixup (as for Relaxation) because we didn't want to emit the R_RISCV_VENDOR relocation if the fixup could be evaluated/applied by the compiler, and that kind of "evaluating one fixup deletes another fixup" seemed complex and error-prone.

MaskRay replied:
You could customize `evaluateTargetFixup`. If the fixup is not resolved, call `recordRelocation` to append the marker relocation, and then rely on the default `recordRelocation` in `evaluateFixup` to append the actual relocation.

lenary replied:
I looked into this, and `evaluateTargetFixup` doesn't get any info about the fixup being resolved (the info comes from this function when the IsTarget fixup flag is given) - to work this way we'd need to pass more information into this function (the `RecordReloc` bool), and replicate lots of the logic in `evaluateFixup` which is bypassed if you have the `IsTarget` fixup flag.

With all this in mind, I continue to think that the new hook is the best way to achieve vendor relocations for RISC-V.

I realise you've been away recently (me too), I'm happy to discuss this further when you're back.

lenary replied:
@MaskRay Ping?

lenary replied:
I have made the change, and commented about the duplicated code, which I don't like having to copy-paste. There is also complex logic working out when `IsResolved` might not be enough of a signal that a relocation is required, especially as RISC-V forces relocations during relaxation (which is also hard to test).

MaskRay replied:
Cc @jrtc27  who introduced `evaluateTargetFixup`

Currently, the MCFixupKindInfo::FKF_IsTarget code path in MCAssembler.cpp skips the common code paths (getAddSym/getSubSym), because PCREL_LO12[IS] don't need them.

Now evaluateTargetFixup is reused by QUIC relocations, leading to some duplication. I think we can make evaluateTargetFixup return more information so that  QUIC relocations keep reusing the common code path.

```cpp
  if (FixupFlags & MCFixupKindInfo::FKF_IsTarget) {
    IsResolved =
        getBackend().evaluateTargetFixup(*this, Fixup, DF, Target, STI, Value);
  } else {
    const MCSymbol *Add = Target.getAddSym(); 7 refs
    const MCSymbol *Sub = Target.getSubSym();
```

The latter half of the duplication that adds a R_RISCV_VENDOR relocation can probably be refactored to reuse `Type = TargetObjectWriter->getRelocType(Ctx, Target, Fixup, IsPCRel);`

jrtc27 replied:
evaluteTargetFixup did exactly what was needed at the time and nothing more. If there are cases now where more of the surrounding code needs reusing then coming up with a way to communicate that seems perfectly reasonable to me.

MaskRay replied:
I've refactored the relocation generation further. There is now an addReloc hook and  #140494 demonstrates how to add two relocations from one fixup. For VENDOR, we just need to call .recordRelocation for a specific fixup kind.

MaskRay commented on llvm/test/MC/RISCV/xqcibi-relocations.s:
```diff
@@ -34,6 +45,9 @@ qc.e.beqi x7, 8, .L1
 # FIXUP: fixup A - offset: 0, value: .L1, kind: fixup_riscv_qc_e_branch
 
+# Check that there is only one vendor symbol created and that it is local and NOTYPE
+# VENDORSYM-COUNT-1: 00000000     0 NOTYPE  LOCAL  DEFAULT     2 QUALCOMM
```
COUNT-1 doesn't check that the next line does not match the pattern. a NOT is needed

svs-quic replied:
Added a NOT check.

MaskRay commented on llvm/test/MC/RISCV/xqcili-relocations.s:
```diff
@@ -2,31 +2,37 @@
 
 qc.e.li x5, foo
+# RELOC: R_RISCV_VENDOR QUALCOMM 0x0
 # RELOC: R_RISCV_CUSTOM194 foo 0x0
```
`-NEXT` is needed to check that the two relocations are consecutive

svs-quic replied:
Done

lenary commented:
@tclin914 I've tagged you because you'll likely want this functionality for the andes extension with relocations, and it would be good to avoid duplicating this work.

MaskRay commented on llvm/test/MC/RISCV/xqcibi-relocations.s:
```diff
@@ -19,6 +21,15 @@ qc.e.bgeui x8, 12, foo
+# Check that we print the correct relocations in exact mode
+.option exact
+qc.e.bgeui x8, 12, foo
+# RELOC: R_RISCV_VENDOR QUALCOMM 0x0
```
For relocation generation testing we should use llvm-objdump -dr to test r_offset and test that we don't incorrectly emit redundant relocations. See Sparc/Relocations/relocation-specifier.s for a modern example using llvm-objdump -dr

Looks like R_RISCV_VENDOR is only generated with `exact`. We should test that without `exact` there is no `QUALCOMM`. This is unclear from the PR description.

lenary replied:
QUALCOMM should be emitted even without `exact`, we have started to use exact to avoid instruction compression happening and make testing more specific.

I will move to `objdump -dr`, I agree it's often clearer. I will revisit these tests.

resistor commented:
Just wanted to note that we (CHERI, and specifically CHERIoT) are very interested in seeing `R_RISCV_VENDOR` support landed. Is this still moving ahead?

lenary commented:
@resistor I'm going to try to make some more progress on this today.

lenary committed:
[RISCV] Vendor Relocations for Xqci extension

This patch implements vendor relocation support for RISC-V, starting
with the Xqci extensions.

This is done by re-organising `RISCVAsmBackend::evaluateTargetFixup`, to
also handle emitting the vendor relocation when the fixup is unresolved.

An extra parameter `bool RecordReloc` is added to `evaluateTargetFixup`,
so that the relocation is only emitted when it should be, as
`evaluateTargetFixup` works the same way.

This adds tests showing resolveable and unresolvable symbols for the
xqci fixups (some of which are absolute and some of which are
pc-relative), including tests when relaxation is enabled which forces
relocations that might otherwise not be needed.

Co-authored-by: Sudharsan Veeravalli <quic_svs@quicinc.com>

lenary committed:
clang-format

MaskRay commented on llvm/test/MC/RISCV/xqcilb-relocations.s:
```diff
@@ -1,46 +1,72 @@
-
-.text
+## This test checks that we emit the right relocations for Xqcilb
+## relative jumps. These can be resolved within the same section
```
These can be resolved within the same section and the destination symbol is local?

Otherwise I think there will be a relocation as well. (We also emit a relocation when the destination is a gnu ifunc symbol, which is an edge case that we don't necessarily describe).

lenary replied:
Yeah, I think these always need a relocation. I forgot these are to the PLT, was just focussed on them being pc-relative.

I don't think I found the LLVM code to always emit the relocation (that I expect for call_plt), i will look over it all again.

lenary replied:
I added tests to show that for non-local symbols, we still get the relocation, but for local symbols we don't (which matches when `call` is resolved directly to the symbol in the assembler)

MaskRay replied:
You might want to edit one of the tests to define `QUALCOMM:` and use llvm-objdump -t to check the symbol table. There will be two `QUALCOMM` local symbols, which is fine.

MaskRay commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVAsmBackend.cpp:
```diff
@@ -622,6 +598,129 @@ bool RISCVAsmBackend::evaluateTargetFixup(
+          cast<MCSymbolELF>(Ctx.createLocalSymbol(VendorIdentifier));
+
+      // Vendor Symbols are Absolute, Local, NOTYPE.
+      VendorSymbol->setType(ELF::STT_NOTYPE);
```
The symbol is by default NOTYPE and LOCAL. In the rare scenario that the users changes the type/binding, we should respect that.

So setType/setBinding should be omitted.

lenary replied:
Done.

lenary committed:
Merge branch 'main' into vendorsym

MaskRay commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVAsmBackend.cpp:
```diff
@@ -660,6 +699,14 @@ bool RISCVAsmBackend::addReloc(const MCFragment &F, const MCFixup &Fixup,
+
+  // Some Fixups require a vendor relocation, record it (directly) before we add
+  // the relocation.
+  if (!IsResolved || shouldForceRelocation(Fixup, Target)) {
```
If we are going to call `shouldForceRelocation`, just avoid the generic `MCAsmBackend::addReloc` and call the underlying `recordRelocation` directly.

MaskRay commented on llvm/test/MC/RISCV/xqcilb-relocations.s:
```diff
@@ -1,46 +1,92 @@
+qc.e.j same_section
+
+# ASM: qc.e.jal same_section
+# OBJ: qc.e.jal 0x30
```
If we use OBJ-NEXT:, OBJ-NOT: R_RISCV will not be needed

MaskRay commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVAsmBackend.cpp:
```diff
@@ -611,6 +612,44 @@ bool RISCVAsmBackend::evaluateTargetFixup(const MCFixup &Fixup,
+  const MCExpr *VendorExpr = MCSymbolRefExpr::create(VendorSymbol, Ctx);
+  MCFixup VendorFixup =
+      MCFixup::create(Fixup.getOffset(), VendorExpr,
+                      FirstLiteralRelocationKind + ELF::R_RISCV_VENDOR);
```
s/FirstLiteralRelocationKind + ELF::R_RISCV_VENDOR/ELF::R_RISCV_VENDOR/`

lenary committed:
Merge branch 'main' into vendorsym

lenary committed:
Address Review Feedback

lenary commented:
Failure is unrelated.

MaskRay commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVAsmBackend.h:
```diff
@@ -20,6 +21,7 @@ namespace llvm {
 class MCAssembler;
 class MCObjectTargetWriter;
 class raw_ostream;
+class MCSymbolELF;
```
The header file doesn't use MCSymbolELF

MaskRay commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVAsmBackend.h:
```diff
@@ -47,9 +51,15 @@ class RISCVAsmBackend : public MCAsmBackend {
                            uint64_t &Value) override;
 
+  std::optional<StringRef>
+  getVendorIdentifierForFixup(unsigned TargetFixupKind) const;
```
rename this to `FixupKind` to be consistent with the impl.

(Beside FK_Data_ fixup kinds are almost always target-specific. "Target" probably does not convey extra meaning. )

lenary committed:
Address review comments

lenary commented:
The linux build failure is unrelated, it's an issue with precompiled headers in Flang. I will file a separate ticket about it if there isn't one.

lenary committed:
Merge branch 'main' into vendorsym

svs-quic commented:
Thanks for taking this up and working on it, Sam! The changes look good to me.

MaskRay commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVAsmBackend.cpp:
```diff
@@ -38,7 +39,7 @@ static cl::opt<bool> ULEB128Reloc(
                                  bool Is64Bit, const MCTargetOptions &Options)
     : MCAsmBackend(llvm::endianness::little), STI(STI), OSABI(OSABI),
-      Is64Bit(Is64Bit), TargetOptions(Options) {
+      Is64Bit(Is64Bit), TargetOptions(Options), VendorSymbols() {
```
delete `, VendorSymbols()`. just rely the implicit default initialization

MaskRay commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVAsmBackend.cpp:
```diff
@@ -660,7 +698,16 @@ bool RISCVAsmBackend::addReloc(const MCFragment &F, const MCFixup &Fixup,
+  if (!IsResolved) {
+    // Some Fixups require a vendor relocation, record it (directly) before we
+    // add the relocation.
+    if (std::optional<StringRef> VendorIdentifier =
```
Combine the two functions and just call sth like `maybeAddVendorReloc(F, Fixup)`

MaskRay commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVAsmBackend.cpp:
```diff
@@ -611,6 +612,43 @@ bool RISCVAsmBackend::evaluateTargetFixup(const MCFixup &Fixup,
+                                     StringRef VendorIdentifier) {
+  MCContext &Ctx = Asm->getContext();
+
+  auto It = VendorSymbols.find(VendorIdentifier);
```
find/try_emplace performs the hash table lookup twice. Just call try_emplace and use the second element of the return value.

MaskRay commented on llvm/test/MC/RISCV/xqcilb-relocations.s:
```diff
@@ -1,46 +1,84 @@
+# OBJ: qc.e.j 0x0 <this_section>
+# OBJ-NEXT: R_RISCV_VENDOR QUALCOMM{{$}}
+# OBJ-NEXT: R_RISCV_CUSTOM195 undef{{$}}
+qc.e.j undef
```
In the absence of `.option exact`, `qc.e.j undef` will be converted to a shorter-range jal, is that intended?

lenary replied:
Addressed in #142702

lenary committed:
Merge branch 'main' into vendorsym

lenary committed:
Address feedback

lenary committed:
clang-format

MaskRay commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVAsmBackend.cpp:
```diff
@@ -611,6 +612,50 @@ bool RISCVAsmBackend::evaluateTargetFixup(const MCFixup &Fixup,
+void RISCVAsmBackend::maybeAddVendorReloc(const MCFragment &F,
+                                          const MCFixup &Fixup) {
+  MCContext &Ctx = Asm->getContext();
+
```
I'd delete the first two blank lines.

lenary replied:
Done, with a little reorder to not grab the context until as late as possible

MaskRay commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVAsmBackend.cpp:
```diff
@@ -611,6 +612,50 @@ bool RISCVAsmBackend::evaluateTargetFixup(const MCFixup &Fixup,
+
+  const MCExpr *VendorExpr = MCSymbolRefExpr::create(VendorSymbol, Ctx);
+  MCFixup VendorFixup =
+      MCFixup::create(Fixup.getOffset(), VendorExpr, ELF::R_RISCV_VENDOR);
```
`MCFixup::create(Fixup.getOffset(), nullptr, ELF::R_RISCV_VENDOR)` Then delete MCSymbolRefExpr::create .  MCFixup::Value isn't used

lenary replied:
Done

MaskRay commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVAsmBackend.cpp:
```diff
@@ -611,6 +612,50 @@ bool RISCVAsmBackend::evaluateTargetFixup(const MCFixup &Fixup,
+    Asm->registerSymbol(*VendorSymbol);
+  } else {
+    // Fetch the existing symbol
+    VendorSymbol = It->getValue();
```
the objdump -t CHECK lines need some -NOT pattern to check that we don't generate redundat local symbols.

lenary replied:
Done

lenary committed:
Address review feedback

lenary committed:
Merge branch 'main' into vendorsym

lenary reviewed (commented):
This has been a long time in review, so I'm probably going to merge it fairly early tomorrow (pacific time) if there are no more comments.
